Common/src/gRPC/Services/WatchToGrpc.cs

```diff
@@ -0,0 +1,299 @@
+// but WITHOUT ANY WARRANTY, without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+// 
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+using ArmoniK.Api.gRPC.V1.Events;
+using ArmoniK.Core.Common.Storage;
+using ArmoniK.Core.Common.Storage.Events;
+
+using Microsoft.Extensions.Logging;
+
+namespace ArmoniK.Core.Common.gRPC.Services;
+
+public class WatchToGrpc
+{
+  private const    int            PageSize = 100;
+  private readonly ILogger        logger_;
+  private readonly IResultTable   resultTable_;
+  private readonly IResultWatcher resultWatcher_;
+  private readonly ITaskTable     taskTable_;
+  private readonly ITaskWatcher   taskWatcher_;
+
+  public WatchToGrpc(ITaskTable     taskTable,
+                     ITaskWatcher   taskWatcher,
+                     IResultTable   resultTable,
+                     IResultWatcher resultWatcher,
+                     ILogger        logger)
+  {
+    taskTable_     = taskTable;
+    taskWatcher_   = taskWatcher;
+    resultTable_   = resultTable;
+    resultWatcher_ = resultWatcher;
+    logger_        = logger;
+  }
+
+  public IAsyncEnumerable<EventSubscriptionResponse> GetEvents(string            sessionId,
+                                                               CancellationToken cancellationToken)
+  {
+    var channel = Channel.CreateUnbounded<EventSubscriptionResponse>();
+
+    Task.Factory.StartNew(async () =>
+                          {
+                            var                                           read = 0;
+                            var                                           page = 0;
+                            (IEnumerable<TaskData> tasks, int totalCount) res;
+                            while ((res = await taskTable_.ListTasksAsync(data => data.SessionId == sessionId,
+                                                                          data => data.CreationDate,
+                                                                          false,
+                                                                          page,
+                                                                          PageSize,
+                                                                          cancellationToken)
+                                                          .ConfigureAwait(false)).totalCount > read)
+                            {
+                              using var enumerator = res.tasks.GetEnumerator();
+
+                              while (enumerator.MoveNext())
+                              {
+                                var cur = enumerator.Current;
```

Comment:
I don't think you need to explicitly use the enumerator here:
```suggestion
                              foreach (var cur in res.tasks) {
```